# Imports

In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import plotly.express as px
import matplotlib.style as style

import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows', None)

In [2]:
df = pd.read_csv('data.csv')
genre_df = pd.read_csv('data_by_genres.csv')
year_df = pd.read_csv('data_by_year.csv')
df_w_genre = pd.read_csv('data_w_genres.csv')

In [3]:
df.head()

valence  year  acousticness  \
0   0.0594  1921         0.982   
1   0.9630  1921         0.732   
2   0.0394  1921         0.961   
3   0.1650  1921         0.967   
4   0.2530  1921         0.957   

                                             artists  danceability  \
0  ['Sergei Rachmaninoff', 'James Levine', 'Berli...         0.279   
1                                     ['Dennis Day']         0.819   
2  ['KHP Kridhamardawa Karaton Ngayogyakarta Hadi...         0.328   
3                                   ['Frank Parker']         0.275   
4                                     ['Phil Regan']         0.418   

   duration_ms  energy  explicit                      id  instrumentalness  \
0       831667   0.211         0  4BJqT0PrAfrxzMOxytFOIz          0.878000   
1       180533   0.341         0  7xPhfUan2yNtyFG0cUWkt8          0.000000   
2       500062   0.166         0  1o6I8BglA6ylDMrIELygv1          0.913000   
3       210000   0.309         0  3ftBPsC5vPBKxYSee08FDH          0.000028   
4       166693   0.193         0  4d6HGyGT8e121BsdKmw9v6          0.000002   

   key  liveness  loudness  mode  \
0   10     0.665   -20.096     1   
1    7     0.160   -12.441     1   
2    3     0.101   -14.850     1   
3    5     0.381    -9.316     1   
4    3     0.229   -10.096     1   

                                                name  popularity release_date  \
0  Piano Concerto No. 3 in D Minor, Op. 30: III. ...           4         1921   
1                            Clancy Lowered the Boom           5         1921   
2                                          Gati Bali           5         1921   
3                                          Danny Boy           3         1921   
4                        When Irish Eyes Are Smiling           2         1921   

   speechiness    tempo  
0       0.0366   80.954  
1       0.4150   60.936  
2       0.0339  110.339  
3       0.0354  100.109  
4       0.0380  101.665

In [4]:
genre_df.head()

mode                  genres  acousticness  danceability   duration_ms  \
0     1  21st century classical      0.979333      0.162883  1.602977e+05   
1     1                   432hz      0.494780      0.299333  1.048887e+06   
2     1                   8-bit      0.762000      0.712000  1.151770e+05   
3     1                      []      0.651417      0.529093  2.328809e+05   
4     1              a cappella      0.676557      0.538961  1.906285e+05   

     energy  instrumentalness  liveness   loudness  speechiness       tempo  \
0  0.071317          0.606834  0.361600 -31.514333     0.040567   75.336500   
1  0.450678          0.477762  0.131000 -16.854000     0.076817  120.285667   
2  0.818000          0.876000  0.126000  -9.180000     0.047000  133.444000   
3  0.419146          0.205309  0.218696 -12.288965     0.107872  112.857352   
4  0.316434          0.003003  0.172254 -12.479387     0.082851  112.110362   

    valence  popularity  key  
0  0.103783   27.833333    6  
1  0.221750   52.500000    5  
2  0.975000   48.000000    7  
3  0.513604   20.859882    7  
4  0.448249   45.820071    7

In [5]:
year_df.head()

mode  year  acousticness  danceability    duration_ms    energy  \
0     1  1921      0.886896      0.418597  260537.166667  0.231815   
1     1  1922      0.938592      0.482042  165469.746479  0.237815   
2     1  1923      0.957247      0.577341  177942.362162  0.262406   
3     1  1924      0.940200      0.549894  191046.707627  0.344347   
4     1  1925      0.962607      0.573863  184986.924460  0.278594   

   instrumentalness  liveness   loudness  speechiness       tempo   valence  \
0          0.344878  0.205710 -17.048667     0.073662  101.531493  0.379327   
1          0.434195  0.240720 -19.275282     0.116655  100.884521  0.535549   
2          0.371733  0.227462 -14.129211     0.093949  114.010730  0.625492   
3          0.581701  0.235219 -14.231343     0.092089  120.689572  0.663725   
4          0.418297  0.237668 -14.146414     0.111918  115.521921  0.621929   

   popularity  key  
0    0.653333    2  
1    0.140845   10  
2    5.389189    0  
3    0.661017   10  
4    2.604317    5

In [6]:
df_w_genre.head()

genres                                            artists  \
0  ['show tunes']                   "Cats" 1981 Original London Cast   
1              []                          "Cats" 1983 Broadway Cast   
2              []        "Fiddler On The Roof” Motion Picture Chorus   
3              []     "Fiddler On The Roof” Motion Picture Orchestra   
4              []  "Joseph And The Amazing Technicolor Dreamcoat"...   

   acousticness  danceability    duration_ms    energy  instrumentalness  \
0      0.590111      0.467222  250318.555556  0.394003          0.011400   
1      0.862538      0.441731  287280.000000  0.406808          0.081158   
2      0.856571      0.348286  328920.000000  0.286571          0.024593   
3      0.884926      0.425074  262890.962963  0.245770          0.073587   
4      0.510714      0.467143  270436.142857  0.488286          0.009400   

   liveness   loudness  speechiness       tempo   valence  popularity  key  \
0  0.290833 -14.448000     0.210389  117.518111  0.389500   38.333333    5   
1  0.315215 -10.690000     0.176212  103.044154  0.268865   30.576923    5   
2  0.325786 -15.230714     0.118514   77.375857  0.354857   34.857143    0   
3  0.275481 -15.639370     0.123200   88.667630  0.372030   34.851852    0   
4  0.195000 -10.236714     0.098543  122.835857  0.482286   43.000000    5   

   mode  count  
0     1      9  
1     1     26  
2     1      7  
3     1     27  
4     1      7

In [7]:
dfs = [df,genre_df,year_df,df_w_genre]

for i in dfs:
    print(f'There are {i.duplicated().sum()} duplicated data.')
    print(i.info())
    print('\n')

There are 0 duplicated data.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170653 entries, 0 to 170652
Data columns (total 19 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   valence           170653 non-null  float64
 1   year              170653 non-null  int64  
 2   acousticness      170653 non-null  float64
 3   artists           170653 non-null  object 
 4   danceability      170653 non-null  float64
 5   duration_ms       170653 non-null  int64  
 6   energy            170653 non-null  float64
 7   explicit          170653 non-null  int64  
 8   id                170653 non-null  object 
 9   instrumentalness  170653 non-null  float64
 10  key               170653 non-null  int64  
 11  liveness          170653 non-null  float64
 12  loudness          170653 non-null  float64
 13  mode              170653 non-null  int64  
 14  name              170653 non-null  object 
 15  popularity        170653 non-null  int6

Great! There no null values in the data sets.

# Correct Formatting

There are unwanted characters like square brackets and quotation marks from artists and genres column in the data frames. Let's remove them.

In [8]:
df['artists'] = df['artists'].str.replace(r"\[|\]|'",'',regex=True)
genre_df['genres'] = genre_df['genres'].str.replace(r"\[|\]|'",'',regex=True)
df_w_genre['genres'] = df_w_genre['genres'].str.replace(r"\[|\]|'",'',regex=True)

Now, we have genres with no values.

In [9]:
genre_df[genre_df['genres']=='']

mode genres  acousticness  danceability   duration_ms    energy  \
3     1             0.651417      0.529093  232880.89025  0.419146   

   instrumentalness  liveness   loudness  speechiness       tempo   valence  \
3          0.205309  0.218696 -12.288965     0.107872  112.857352  0.513604   

   popularity  key  
3   20.859882    7

Just one data missing genre, I will just fill it in with 'other'.

In [10]:
genre_df[genre_df['genres']==''] = genre_df[genre_df['genres']==''].replace('','other')

In [11]:
df_w_genre[df_w_genre['genres']==''].head()

genres                                            artists  acousticness  \
1                                 "Cats" 1983 Broadway Cast      0.862538   
2               "Fiddler On The Roof” Motion Picture Chorus      0.856571   
3            "Fiddler On The Roof” Motion Picture Orchestra      0.884926   
4         "Joseph And The Amazing Technicolor Dreamcoat"...      0.510714   
5         "Joseph And The Amazing Technicolor Dreamcoat"...      0.609556   

   danceability    duration_ms    energy  instrumentalness  liveness  \
1      0.441731  287280.000000  0.406808          0.081158  0.315215   
2      0.348286  328920.000000  0.286571          0.024593  0.325786   
3      0.425074  262890.962963  0.245770          0.073587  0.275481   
4      0.467143  270436.142857  0.488286          0.009400  0.195000   
5      0.487278  205091.944444  0.309906          0.004696  0.274767   

    loudness  speechiness       tempo   valence  popularity  key  mode  count  
1 -10.690000     0.176212  103.044154  0.268865   30.576923    5     1     26  
2 -15.230714     0.118514   77.375857  0.354857   34.857143    0     1      7  
3 -15.639370     0.123200   88.667630  0.372030   34.851852    0     1     27  
4 -10.236714     0.098543  122.835857  0.482286   43.000000    5     1      7  
5 -18.266389     0.098022  118.648944  0.441556   32.777778    5     1     36

In [12]:
len(df_w_genre[df_w_genre['genres']==''])

9857

There are quite a lot data missing genres in df_w_genre, for now, let's just fill them with 'other'. Hopefully, it is not affecting the following processes.

In [13]:
df_w_genre[df_w_genre['genres']==''] = df_w_genre[df_w_genre['genres']==''].replace('','other')

In [14]:
len(df_w_genre[df_w_genre['genres']==''])

0

Looks like the year and release date column have the same information, let's check.

In [15]:
print(df['year'].unique())
print(f"There are {df['year'].nunique()} unique years.")

[1921 1922 1923 1924 1925 1926 1927 1928 1929 1930 1931 1932 1933 1934
 1935 1936 1937 1938 1939 1940 1941 1942 1943 1944 1945 1946 1947 1948
 1949 1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962
 1963 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976
 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990
 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004
 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018
 2019 2020]
There are 100 unique years.


In [16]:
print(df['release_date'].unique())
print(f"There are {df['release_date'].nunique()} unique release dates.")

['1921' '1921-03-20' '1921-03-27' ... '2020-04-15' '2020-05-25'
 '2020-11-03']
There are 11244 unique release dates.


There are way more release dates than years, obviously we are going to analyze the musics based on monthly or daily information. I will drop the release date column for it is redundant. But first, let's make sure all the values in the year column are indeed equal to years in release date.

In [17]:
# parse out the year from release date, and convert datatype
df['release_date'] = df['release_date'].apply(lambda x: x.split('-')[0]).astype(int)

In [18]:
# check if two columns are equal
print(f"There are {(len(df))-((df['year']==df['release_date']).sum())} data do not match.")

There are 0 data do not match.


Great! It is now safe to drop one of these columns.

In [19]:
df.drop(columns='release_date',inplace=True)

# Handle Defective Data

In [20]:
# describe the numerical features
df.describe().T

count           mean            std     min          25%  \
valence           170653.0       0.528587       0.263171     0.0       0.3170   
year              170653.0    1976.787241      25.917853  1921.0    1956.0000   
acousticness      170653.0       0.502115       0.376032     0.0       0.1020   
danceability      170653.0       0.537396       0.176138     0.0       0.4150   
duration_ms       170653.0  230948.310666  126118.414668  5108.0  169827.0000   
energy            170653.0       0.482389       0.267646     0.0       0.2550   
explicit          170653.0       0.084575       0.278249     0.0       0.0000   
instrumentalness  170653.0       0.167010       0.313475     0.0       0.0000   
key               170653.0       5.199844       3.515094     0.0       2.0000   
liveness          170653.0       0.205839       0.174805     0.0       0.0988   
loudness          170653.0     -11.467990       5.697943   -60.0     -14.6150   
mode              170653.0       0.706902       0.455184     0.0       0.0000   
popularity        170653.0      31.431794      21.826615     0.0      11.0000   
speechiness       170653.0       0.098393       0.162740     0.0       0.0349   
tempo             170653.0     116.861590      30.708533     0.0      93.4210   

                            50%          75%          max  
valence                0.540000       0.7470        1.000  
year                1977.000000    1999.0000     2020.000  
acousticness           0.516000       0.8930        0.996  
danceability           0.548000       0.6680        0.988  
duration_ms       207467.000000  262400.0000  5403500.000  
energy                 0.471000       0.7030        1.000  
explicit               0.000000       0.0000        1.000  
instrumentalness       0.000216       0.1020        1.000  
key                    5.000000       8.0000       11.000  
liveness               0.136000       0.2610        1.000  
loudness             -10.580000      -7.1830        3.855  
mode                   1.000000       1.0000        1.000  
popularity            33.000000      48.0000      100.000  
speechiness            0.045000       0.0756        0.970  
tempo                114.729000     135.5370      243.507

One thing to notice, most of these attributes are of value in a relative scale, but tempo or beats per minute is actual numerical representation of a song. It is impossible for a song to have 0 tempo, let's impute these values to the average.

In [21]:
# how many songs with 0 tempo
print(f"There are {len(df[df['tempo']==0])} songs with 0 tempo.")

There are 143 songs with 0 tempo.


Let's impute these values to the mean instead.

In [22]:
df.loc[df['tempo']==0,'tempo'] = df.loc[df['tempo']>0,'tempo'].mean()
df_w_genre.loc[df_w_genre['tempo']==0,'tempo'] = df_w_genre.loc[df_w_genre['tempo']>0,'tempo'].mean()

In [23]:
# sanity check
df['tempo'].describe()

count    170653.000000
mean        116.959597
std          30.521479
min          30.946000
25%          93.499000
50%         114.803000
75%         135.537000
max         243.507000
Name: tempo, dtype: float64

# Drop Duplicates

It looks like there are no duplicates in the data frame. But notice the id column in df contains unique information, let's remove this column and check duplicates again.

In [24]:
# check duplicates without id
df = df.drop(columns='id')

In [25]:
# create dictionary for the dataframes
df_list = {'track dataframe':df,
           'year dataframe':year_df,
           'genre dataframe':genre_df,
           'track with genre dataframe':df_w_genre}

# print the number of duplicates
for name, i in df_list.items():
    print(f"There are {i.duplicated().sum()}, {round(i.duplicated().sum()/len(i)*100,4)}% duplicated data in {name}.")

There are 565, 0.3311% duplicated data in track dataframe.
There are 0, 0.0% duplicated data in year dataframe.
There are 0, 0.0% duplicated data in genre dataframe.
There are 0, 0.0% duplicated data in track with genre dataframe.


We have 565 duplicated data in the track dataframe, after excluding the ids, let's look at a duplicated track again to make sure.

In [26]:
df[df.duplicated()].head()

valence  year  acousticness                       artists  \
2298    0.7700  1933        0.9300                Billie Holiday   
2340    0.9450  1933        0.9580  Billie Holiday, Teddy Wilson   
18090   0.0985  2012        0.0152                Kendrick Lamar   
19464   0.1830  2019        0.6040                 Lewis Capaldi   
19715   0.3490  2020        0.2070                     Lil Mosey   

       danceability  duration_ms  energy  explicit  instrumentalness  key  \
2298          0.721       155667   0.186         0          0.000145    3   
2340          0.761       173667   0.420         0          0.037500   10   
18090         0.587       310720   0.698         1          0.000000    1   
19464         0.459       215107   0.575         0          0.000000    3   
19715         0.774       162547   0.554         1          0.000000    0   

       liveness  loudness  mode  \
2298     0.3510   -13.833     1   
2340     0.1390   -10.011     1   
18090    0.4420    -8.948     1   
19464    0.0885    -4.858     1   
19715    0.1320    -7.909     1   

                                                    name  popularity  \
2298                                 No Regrets - Take 1          17   
2340   Spreadin' Rhythm Around (with Teddy Wilson & H...          15   
18090                          Bitch, Don’t Kill My Vibe          66   
19464                                      Before You Go          86   
19715                                    Blueberry Faygo          84   

       speechiness    tempo  
2298        0.0458  129.452  
2340        0.1220   95.492  
18090       0.0943  133.954  
19464       0.0573  111.881  
19715       0.0383   99.034

In [27]:
# look at duplicates of 'Bitch, Don’t Kill My Vibe'
df[(df['artists']=="Kendrick Lamar") & (df['name']=='Bitch, Don’t Kill My Vibe')]

valence  year  acousticness         artists  danceability  duration_ms  \
18079   0.0985  2012        0.0152  Kendrick Lamar         0.587       310720   
18090   0.0985  2012        0.0152  Kendrick Lamar         0.587       310720   

       energy  explicit  instrumentalness  key  liveness  loudness  mode  \
18079   0.698         1               0.0    1     0.442    -8.948     1   
18090   0.698         1               0.0    1     0.442    -8.948     1   

                            name  popularity  speechiness    tempo  
18079  Bitch, Don’t Kill My Vibe          66       0.0943  133.954  
18090  Bitch, Don’t Kill My Vibe          66       0.0943  133.954

It is safe to just drop them.

In [28]:
# drop duplicates
df = df[~df.duplicated()==True]

In [29]:
# check
df.shape

(170088, 17)

# Export CSV

In [30]:
df.to_csv('df_cleaned.csv',index=False)
genre_df.to_csv('genre_df_cleaned.csv',index=False)
year_df.to_csv('year_df_cleaned.csv',index=False)
df_w_genre.to_csv('df_w_genre_cleaned.csv',index=False)